lib/wasix/src/runtime/module_cache/filesystem.rs

```diff
@@ -38,66 +37,120 @@ impl FileSystemCache {
     }
 }
 
+/// Loads a module from the filesystem cache.
+///
+/// A tokio reactor must be available
+#[tracing::instrument(level = "debug", skip_all, fields(? path))]
+async fn tokio_load(path: PathBuf, engine: Engine) -> Result<Module, CacheError> {
+    let bytes = read_file(&path).await?;
+    let deserialized = tokio::task::spawn_blocking(move || deserialize(&bytes, &engine))
+        .await
+        .unwrap();
+    match deserialized {
+        Ok(m) => {
+            tracing::debug!("Cache hit!");
+            Ok(m)
+        }
+        Err(e) => {
+            tracing::debug!(
+                path=%path.display(),
+                error=&e as &dyn std::error::Error,
+                "Deleting the cache file because the artifact couldn't be deserialized",
+            );
+
+            if let Err(e) = std::fs::remove_file(&path) {
+                tracing::warn!(
+                    path=%path.display(),
+                    error=&e as &dyn std::error::Error,
+                    "Unable to remove the corrupted cache file",
+                );
+            }
+            Err(e)
+        }
+    }
+}
+
+/// Checks if the path exists in the filesystem cache.
+///
+/// A tokio reactor must be available
+async fn tokio_contains(path: PathBuf) -> Result<bool, CacheError> {
+    tokio::fs::try_exists(&path)
+        .await
+        .map_err(|e| CacheError::FileRead {
+            path: path.clone(),
+            error: e,
+        })
+}
+
+/// Saves the module to the filesystem cache.
+///
+/// A tokio reactor must be available
+#[tracing::instrument(level = "debug", skip_all, fields(? path))]
+async fn tokio_save(path: PathBuf, module: Module) -> Result<(), CacheError> {
+    let parent = path
+        .parent()
+        .expect("Unreachable - always created by joining onto cache_dir");
+
+    if let Err(e) = tokio::fs::create_dir_all(parent).await {
+        tracing::warn!(
+            dir=%parent.display(),
+            error=&e as &dyn std::error::Error,
+            "Unable to create the cache directory",
+        );
+    }
+
+    // Note: We save to a temporary file and persist() it at the end so
+    // concurrent readers won't see a partially written module.
+    let (file, temp) = NamedTempFile::new_in(parent)
```

Comment:
This is a blocking operation, should be in a blocking function.

You could just leave a TODO.